Add the set_component_names client API

mprahl opened the pull request:
Add an API that can create, update, and/or merge components to ensure one canonical component remains.

Addresses FACTORY-3005

mprahl committed:
Move set_connection directly to the client.query module

mprahl committed:
Set the base property on Component.alternative_names to enforce strings

mprahl committed:
Add the set_component_names client API

Add an API that can create, update, and/or merge components to ensure
one canonical component remains.

Addresses FACTORY-3005

mprpic commented on assayist/client/modify.py:
```diff
@@ -0,0 +1,103 @@
+
+    if alternatives is None:
+        alternatives = []
+    if not isinstance(alternatives, list) and not isinstance(alternatives, tuple):
```
This should probably just check if it's an `abc.Iterable`. `ArrayProperty` constructs a list out of any iterable passed to it under the hood, so sets are also valid for example.

mprahl replied:
Good idea.

mprahl replied:
As discussed in IRC, `collections.abc.Iterable` will be `True` when checking this on a string.

mprpic replied:
One minor improvement might be providing the types as a tuple: `isinstance(alternatives, (tuple, list))`:
> If classinfo is a tuple of class or type objects (or recursively, other such tuples), return true if object is an instance of any of the classes or types. [[ref](https://docs.python.org/3.7/library/functions.html#isinstance)]

mprpic commented on assayist/client/modify.py:
```diff
@@ -0,0 +1,103 @@
+    """.format(c_type, c_namespace, ' OR '.join(names_where_clause))
+
+    results, _ = neomodel.db.cypher_query(query)
+    components = [Component.inflate(row[0]) for row in results]
```
Why not use neomodel for this? E.g.:

```python
Component.nodes.filter(canonical_type=c_type, canonical_namespace=c_namespace, canonical_name__in=names)
```

mprahl replied:
I need to filter on `alternative_names` too, and neomodel doesn't support that filter as far as I could tell. You can read the filter code in https://github.com/neo4j-contrib/neomodel/blob/master/neomodel/match.py#L129.

I had filed https://github.com/neo4j-contrib/neomodel/issues/379 with neomodel on Friday asking for this support.

mprpic commented on assayist/client/modify.py:
```diff
@@ -0,0 +1,103 @@
+        raise ValueError('The alternatives keyword argument must be None, a list, or a tuple')
+
+    for alternative in alternatives:
+        if not isinstance(alternative, str):
```
This is a bit unrelated, but perhaps we should define a `StringProperty()` base property restriction on this field in the model, e.g.:
```python
alternative_names = ArrayProperty(StringProperty(), index=True)
```

mprahl replied:
I already did this in commit e65446fb101da7d9a8c428566f1daeef2a331e6a.

mprpic replied:
Bah, I'm blind.

mprpic commented on assayist/client/modify.py:
```diff
@@ -0,0 +1,103 @@
+
+    :param str c_name: the canonical name of the component
+    :param str c_type: the canonical type of the component
+    :param str c_namespace: the canonical namespace of the component
```
Missing `alternatives` in the doc string.

mprpic commented on assayist/client/modify.py:
```diff
@@ -0,0 +1,103 @@
+        raise ValueError('The canonical name can\'t also be an alternative')
+
+    # Create a new list with the canonical name and alternatives
+    names = [c_name] + alternatives
```
If alternatives is passed as a tuple, this will fail. Better to use `itertools.chain` to concatenate two iterables.

mprahl commented:
Thanks for the review. The comments were addressed and this is ready for review again.

mprpic commented on assayist/client/modify.py:
```diff
@@ -0,0 +1,106 @@
+    :kwarg list alternatives: the alternative names associated with the component
+    :raises ValueError: if the input isn't the proper type or c_name's value is also in alternatives
+    """
+    for arg in (c_name, c_type, c_namespace):
```
Should we also validate that name and type are not empty strings since the model accepts that as a valid value?

mprahl replied:
I thought of that but `c_namespace` can be an empty string. I can add another for loop.

mprpic commented on assayist/client/modify.py:
```diff
@@ -0,0 +1,106 @@
+
+    for c in components:
+        # If this property is null in Neo4j, Neomodel doesn't treat it as an empty list, so just set
+        # it here to make the code below simpler since it can always assume it's a list
```
I guess this can be solved by adding an empyt list as the default value of this attribute:
```python
alternative_names = ArrayProperty(base_property=StringProperty(), index=True, default=list)
```

mprahl replied:
Good idea!

mprpic commented on assayist/client/modify.py:
```diff
@@ -0,0 +1,106 @@
+                    alternative not in component.alternative_names:
+                component.alternative_names.append(alternative)
+        # Delete this component since all its information is now stored in the first component
+        c.delete()
```
As mentioned in today's standup, should we log the component that is being removed so we have an audit trail in case we need to revert some changes?

mprahl replied:
Where would we log that?

mprpic replied:
`log.info()` I guess. That should still be shown in the Jenkins jobs, no?

mprahl replied:
The user would be calling this particular code, and not Jenkins. So I'll just use a print statement instead.

Edit: I'll use a `log.warn` instead.

mprpic replied:
Ah, that's right, I forgot this is part of the client lib. `log.warn` makes sense. Thanks!

mprpic commented on assayist/client/modify.py:
```diff
@@ -0,0 +1,106 @@
+    # If there was only one component returned, then it's all set and no merging is required
+    if len(components) == 1:
+        component.save()
+        return
```
This is not really necessary since if the components list is of size 1, the for-loop below will have zero iterations and will continue to saving the component.

mprahl replied:
I had it there for readability, but I removed it.

mprpic reviewed (approved):
:+1:

mprahl committed:
Set the Component.alternative_names default to an empty list

This makes it so that the rest of the code can always assume the type
is a list instead of it potentially being `None`.

mprahl commented:
@sdherr could you please review this?

mprahl committed:
Add logging to the client

mprahl commented:
@mprpic I added logging as requested in a separate commit. May you please review?

mprpic commented:
> @mprpic I added logging as requested in a separate commit. May you please review?

Looks great!

sdherr commented on assayist/client/modify.py:
```diff
@@ -0,0 +1,104 @@
+            raise ValueError('The alternatives keyword argument must only contain strings')
+
+    if c_name in alternatives:
+        raise ValueError('The canonical name can\'t also be an alternative')
```
Instead of throwing an error this seems like a case that could be silently handled by deleting the alternative entry.

mprahl replied:
I don't have a preference so I'll silently handle it as suggested.

sdherr commented on assayist/client/modify.py:
```diff
@@ -0,0 +1,104 @@
+    """.format(c_type, c_namespace, ' OR '.join(names_where_clause))
+
+    results, _ = neomodel.db.cypher_query(query)
+    components = [Component.inflate(row[0]) for row in results]
```
As a personal preference, I would prefer if the query building and inflating logic lived in a Component class method.

mprahl replied:
I think that's a good idea in general, but this query seems very specific to this use-case. A more generic one to find a component based on a name (canonical name or alternative name), namespace, and type would be very useful, and should be done part of the work to take into account the alternative names in the main analyzer.

sdherr commented on assayist/client/modify.py:
```diff
@@ -0,0 +1,104 @@
+        for alternative in c.alternative_names:
+            if alternative != component.canonical_name and \
+                    alternative not in component.alternative_names:
+                component.alternative_names.append(alternative)
```
All of the altnerative_names loops and if statements in the last 40ish lines can probably be more simply handled by:
1) Emptying component.alternative_names
2) Adding the desired entries to a set
3) Iteratively calling component.alternative_names.append at the very end.

mprahl replied:
Very nice suggestion.

sdherr reviewed (approved):
Some minor comments, but I'm fine with it as-is.

mprahl committed:
Simplify the logic in set_component_names

sdherr commented on assayist/client/modify.py:
```diff
@@ -0,0 +1,92 @@
+    all_alt_names = set(alternatives)
+    for c in components:
+        all_alt_names.add(c.canonical_name)
+        all_alt_names = all_alt_names | set(c.alternative_names)
```
To me `all_alt_names.update(c.alternative_names)` is more readable, but they're equivalent.

sdherr commented on assayist/client/modify.py:
```diff
@@ -0,0 +1,92 @@
+
+    # By removing the correct canonical name, we have a set of the correct alternative names
+    if c_name in all_alt_names:
+        all_alt_names.remove(c_name)
```
These two lines can be replaced with `all_alt_names.discard(c_name)`

sdherr reviewed (approved):
Two extremely minor set comments, but looks good to me.
